File: board.py
from typing import List, Tuple

class Board(object):
    def __init__(self, board=None):
        # 0 = empty, 1 = occupied, 2 = invalid
        self.board = board or [
            [2, 2, 1, 1, 1, 2, 2],
            [2, 2, 1, 1, 1, 2, 2],
            [1, 1, 1, 1, 1, 1, 1],
            [1, 1, 1, 0, 1, 1, 1],
            [1, 1, 1, 1, 1, 1, 1],
            [2, 2, 1, 1, 1, 2, 2],
            [2, 2, 1, 1, 1, 2, 2],
        ]

    def __hash__(self):
        return hash(tuple([tuple(row) for row in self.board]))

    def clone(self):
        # Ca. 2x faster than copy.deepcopy()
        board_copy = [[peg for peg in row] for row in self.board]
        return Board(board_copy)

    def score(self) -> int:
        # Count all pegs
        s = 0
        for peg in [p for row in self.board for p in row]:
            s += 1 if peg == 1 else 0

        # If there is only one peg left and it is in the center, the score is 0.
        if s == 1 and self.board[3][3] == 1:
            s = 0

        return s

    def possible_moves(self) -> List[Tuple[Tuple[int, int], Tuple[int, int]]]:
        moves = []  # Format: ((from x, from y), (to x, to y))

        # For each board position
        for x, row in enumerate(self.board):
            for y, peg in enumerate(row):
                # If occupied by a peg
                if peg == 1:
                    # Find valid moves for this peg
                    peg_moves = self.moves_for_peg(x, y)

                    moves.extend([((x, y), move) for move in peg_moves])

        return moves

    def moves_for_peg(self, x, y) -> List[Tuple[int, int]]:
        assert 0 <= x <= 6
        assert 0 <= y <= 6

        peg_moves = []

        # x, y offsets for moves towards top, bottom, left, and right
        move_offsets = [(-2, 0), (2, 0), (0, -2), (0, 2)]

        for (dx, dy) in move_offsets:
            new_x = x + dx
            new_y = y + dy

            # If new position is inside the board
            if 0 <= new_x <= 6 and 0 <= new_y <= 6:
                # If the new position is free
                if self.board[new_x][new_y] == 0:
                    # If there is a peg next to the current peg in the move's direction
                    if self.board[(x + new_x) // 2][(y + new_y) // 2] == 1:
                        peg_moves.append((new_x, new_y))

        return peg_moves

    def move(self, move: Tuple[Tuple[int, int], Tuple[int, int]]):
        (from_x, from_y), (to_x, to_y) = move

        # Delete peg from old position
        assert self.board[from_x][from_y] == 1
        self.board[from_x][from_y] = 0
        # Place peg at new position
        assert self.board[to_x][to_y] == 0
        self.board[to_x][to_y] = 1
        # Delete peg in between
        assert self.board[(from_x + to_x) // 2][(from_y + to_y) // 2] == 1
        self.board[(from_x + to_x) // 2][(from_y + to_y) // 2] = 0

        return self

    def __repr__(self):
        s = []
        for j in range(len(self.board[0])):
            for i in range(len(self.board)):
                x = self.board[i][j]
                if x == 1:
                    s += "+"
                else:
                    s += ' '
                #s += str(self.board[i][j])
            s += '\n'

        return ''.join(s)

class IterativeSolver:
    def __init__(self, board):
        # Set of hashes of board positions. Used to skip boards that have been played already.
        self.boards_visited = set()
        self.board = board
        self.stack = [board]
        self.parent = {}

    def build_solution(self):
        solution = []
        while True:
            try:
                move,self.board = self.parent[self.board]
                solution.append(move)
            except KeyError:
                return list(reversed(solution))
            
    def solve_internal(self):
        self.board = self.stack.pop()
        if hash(self.board) not in self.boards_visited:
            self.boards_visited.add(hash(self.board))

            moves = self.board.possible_moves()
            if len(moves) == 0:
                score = self.board.score()
                if score == 0:
                    solution = self.build_solution()
                    return solution
                    #print(self.parent)
                    
            for move in moves:
                b = self.board.clone().move(move)
                self.parent[b] = move, self.board
                self.stack.append(b)
    
    def solve(self):
        while True:
            result = self.solve_internal()
            if result:
                return result

class RecursiveSolver:
    def __init__(self, move_callback=None, done_callback=None):
        # Set of hashes of board positions. Used to skip boards that have been played already.
        self.boards_played = set()

        # Counters for statistical purposes.
        self.statistics = {'Games finished': 0, 'Boards skipped': 0}
        self.move_callback = move_callback
        self.done_callback = done_callback

    def solve(self, board, move_memo=()):
        if self.move_callback:
            self.move_callback(board, move_memo)

        if hash(board) in self.boards_played:
            self.statistics['Boards skipped'] += 1
            return
        self.boards_played.add(hash(board))
        
        moves = board.possible_moves()

        # If there are no moves left
        if len(moves) == 0:
            self.statistics['Games finished'] += 1
            if self.done_callback:
                self.done_callback(board, move_memo)
            # If the game is solved
            if board.score() == 0:
                return move_memo
        else:
            for move in moves:
                result = self.solve_recursive(board.clone().move(move), [mm for mm in move_memo] + [move])
                if result:
                    return result

if __name__ == '__main__':
    b = Board()
    s = IterativeSolver(b)
    print(s.solve())
    
    # s = RecursiveSolver()
    # while len(s.stack):
    #     result = s.solve_iterative()
    #     if result:
    #         break

    # print(list(result))

File: board.cpp
#include "board.h"
#include <vector>
#include <optional>
#include <unordered_set>
#include <unordered_map>
#include <stack>
#include <iostream>
#include <list>
#include <algorithm>

Board::Board()
{
}
Board::Board(const Board &board)
{
    std::copy(&board.board_[0][0], &board.board_[0][0] + nx * ny, &board_[0][0]);
}
const int Board::board_at(int i, int j)
{
    return board_[i][j];
}

int Board::score()
{
    int s = 0;
    for (auto i = 0; i < nx; i++)
    {
        for (auto j = 0; j < ny; j++)
        {
            if (board_[i][j] == 1)
                s++;
        }
    }
    if (s == 1 && board_[3][3] == 1)
        s = 0;

    return s;
}

moves_t Board::possible_moves()
{
    moves_t moves;
    for (auto x = 0; x < nx; x++)
    {
        for (auto y = 0; y < ny; y++)
        {
            if (board_[x][y] == 1)
            {
                auto peg_moves = moves_for_peg(x, y);
                for (auto peg_move : peg_moves)
                    moves.push_back(std::make_pair(std::make_pair(x, y), peg_move));
            }
        }
    }
    return moves;
}

peg_moves_t Board::moves_for_peg(int x, int y)
{
    peg_moves_t peg_moves;

    peg_moves_t move_offsets = {{-2, 0}, {2, 0}, {0, -2}, {0, 2}};
    for (auto dxdy : move_offsets)
    {
        int new_x = x + dxdy.first;
        int new_y = y + dxdy.second;
        if (0 <= new_x && new_x <= 6 && 0 <= new_y && new_y <= 6)
        {
            if (board_[new_x][new_y] == 0)
            {
                if (board_[(x + new_x) / 2][(y + new_y) / 2] == 1)
                {
                    peg_moves.push_back(position_t(new_x, new_y));
                }
            }
        }
    }
    return peg_moves;
}

void Board::move(move_t move)
{
    auto fxy = move.first;
    auto txy = move.second;
    auto from_x = fxy.first;
    auto from_y = fxy.second;
    auto to_x = txy.first;
    auto to_y = txy.second;

    board_[from_x][from_y] = 0;
    board_[to_x][to_y] = 1;
    board_[(from_x + to_x) / 2][(from_y + to_y) / 2] = 0;
}

std::ostream &operator<<(std::ostream &out, const Board& board)
{
    for (auto i = 0; i < nx; i++)
    {
        for (auto j = 0; j < ny; j++)
        {
            std::cout << board.board_[i][j];
        }
        std::cout << std::endl;
    }
    return std::cout;
}

bool Board::operator==(const Board board) const
{
    for (int i = 0; i < nx; i++)
    {
        for (int j = 0; j < ny; j++)
        {
            if (board_[i][j] != board.board_[i][j])
                return false;
        }
    }
    return true;
}

Solver::Solver(Board board) : board_(board)
{
    stack.push(board);
}

moves_t Solver::build_solution()
{
    moves_t solution;
    while (true)
    {
        if (!parent.contains(board_))
        {
            solution.reverse();
            return solution;
        }
        auto move_and_board = parent[board_];
        solution.push_back(move_and_board.first);
        board_ = move_and_board.second;
    }
}

moves_t Solver::solve_internal()
{
    board_ = stack.top();
    stack.pop();
    if (!boards_visited.contains(board_))
    {
        boards_visited.insert(board_);
        auto moves = board_.possible_moves();
        if (moves.size() == 0)
        {
            auto score = board_.score();
            if (score == 0) {
                return build_solution();
            }
        }
        for (auto move : moves)
        {
            Board b(board_);
            b.move(move);
            auto p = make_pair(move, board_);
            auto i = make_pair(b, p);
            parent.insert(i);
            stack.push(b);
        }
    }
    return moves_t();
}

void Solver::solve()
{
    moves_t result;
    while (true)
    {
        result = solve_internal();
        if (result.size())
            break;
    }
    for(auto move: result) {
        std::cout << "Move: " << move.first.first << "," << move.first.second << " " << move.second.first << "," << move.second.second << std::endl;
    }
}

int main(void)
{

    Board b;
    Solver s(b);
    s.solve();
}